In [1]:
# 라이브러리 불러오기
import numpy as np
import pandas as pd
from sklearn import metrics
from sklearn import tree
from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsClassifier

# 1번

In [2]:
# load csv
df = pd.read_csv('Housing.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     486 non-null    float64
 1   ZN       486 non-null    float64
 2   INDUS    486 non-null    float64
 3   CHAS     486 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      486 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    int64  
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    486 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB


In [4]:
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622    3  222     18.7   

        B  LSTAT  MEDV  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90    NaN  36.2

In [5]:
# check NaN
df.isnull().sum()

CRIM       20
ZN         20
INDUS      20
CHAS       20
NOX         0
RM          0
AGE        20
DIS         0
RAD         0
TAX         0
PTRATIO     0
B           0
LSTAT      20
MEDV        0
dtype: int64

In [6]:
df.dropna(subset=['CRIM', 'ZN', 'INDUS', 'CHAS', 'AGE', 'LSTAT'],how='any', axis = 0, inplace=True)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 394 entries, 0 to 504
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     394 non-null    float64
 1   ZN       394 non-null    float64
 2   INDUS    394 non-null    float64
 3   CHAS     394 non-null    float64
 4   NOX      394 non-null    float64
 5   RM       394 non-null    float64
 6   AGE      394 non-null    float64
 7   DIS      394 non-null    float64
 8   RAD      394 non-null    int64  
 9   TAX      394 non-null    int64  
 10  PTRATIO  394 non-null    float64
 11  B        394 non-null    float64
 12  LSTAT    394 non-null    float64
 13  MEDV     394 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 46.2 KB


In [8]:
# 1번 케이스
ndf = df
x = ndf[['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT']]
y = ndf[['MEDV']]

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.1, random_state=10)

lr = LinearRegression()
lr.fit(x_train, y_train)

r_square = lr.score(x_test, y_test)
print('r square : ', r_square)
print('계수 : ', lr.coef_)
print('상수항 : ', lr.intercept_)

r square :  0.7687417495006998
계수 :  [[-9.80789002e-02  5.35658388e-02  6.35599321e-02  2.85848512e+00
  -1.99677182e+01  4.12459515e+00 -1.75774036e-02 -1.54025590e+00
   3.15678163e-01 -1.45012117e-02 -9.64786245e-01  9.22806341e-03
  -4.28519193e-01]]
상수항 :  [36.49612327]


In [9]:
# 2번 케이스
ndf = df
x = ndf[[ 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO',  'LSTAT']]
y = ndf[['MEDV']]

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.1, random_state=10)

lr = LinearRegression()
lr.fit(x_train, y_train)

r_square = lr.score(x_test, y_test)
print('r square : ', r_square)
print('계수 : ', lr.coef_)
print('상수항 : ', lr.intercept_)

r square :  0.7583311679135243
계수 :  [[ 3.49390472e+00 -2.05219510e+01  4.15487330e+00 -1.90853223e-02
  -1.19951220e+00  1.94215199e-01 -9.48618337e-03 -1.11526884e+00
  -4.80732587e-01]]
상수항 :  [42.18375849]


In [10]:
# 3번 케이스
ndf = df
x = ndf[['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'B', 'LSTAT']]
y = ndf[['MEDV']]

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.1, random_state=20)

lr = LinearRegression()
lr.fit(x_train, y_train)

r_square = lr.score(x_test, y_test)
print('r square : ', r_square)
print('계수 : ', lr.coef_)
print('상수항 : ', lr.intercept_)

r square :  0.7526615405063615
계수 :  [[-0.09511407  0.07963809 -0.04038044  3.48039808 -7.93323046  4.71916821
  -0.01939042 -1.55736816  0.19329124 -0.01270359  0.00914057 -0.48108251]]
상수항 :  [10.29703225]


## 결과 설명
- 총 3개의 케이스로 나누어 테스트에 들어가는 크기의 데이터는 10%로 설정함.
- 1번 : x->  ndf의 'CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT' 으로 x 값을 정해 회귀분석을 돌린 결과 r squared 0.7687 도출
- 2번 : x -> ndf의 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO',  'LSTAT' 으로 x값을 정해 회귀분석을 돌린 결과 r squared 0.7583 도출
- 3번 : x -> ndf의 'CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'B', 'LSTAT' 으로 x 값을 정해 회귀분석을 돌린 결과 r squared 0.7526 도출

In [11]:
df2 = pd.read_csv('wine.csv')

In [12]:
df2.head(10)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   
5            7.4              0.66         0.00             1.8      0.075   
6            7.9              0.60         0.06             1.6      0.069   
7            7.3              0.65         0.00             1.2      0.065   
8            7.8              0.58         0.02             2.0      0.073   
9            7.5              0.50         0.36             6.1      0.071   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   
5                 13.0                  40.0   0.9978  3.51       0.56   
6                 15.0                  59.0   0.9964  3.30       0.46   
7                 15.0                  21.0   0.9946  3.39       0.47   
8                  9.0                  18.0   0.9968  3.36       0.57   
9                 17.0                 102.0   0.9978  3.35       0.80   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5  
5      9.4        5  
6      9.4        5  
7     10.0        7  
8      9.5        7  
9     10.5        5

In [13]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [14]:
# 결측치 확인 -> 없
df2.isnull().sum()

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

In [15]:
# quality 확인
df2.quality.value_counts()

5    681
6    638
7    199
4     53
8     18
3     10
Name: quality, dtype: int64

In [16]:
# quality 5,6 제외 전체 삭제
mask = df2['quality'].isin([3,4 , 7, 8])
df3 = df2[~mask]

In [17]:
df3

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
0         9.4        5  
1         9.8        5  
2         9.8        5  
3         9.8        6  
4         9.4        5  
...       ...      ...  
1594     10.5        5  
1595     11.2        6  
1596     11.0        6  
1597     10.2        5  
1598     11.0        6  

[1319 rows x 12 columns]

In [18]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1319 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1319 non-null   float64
 1   volatile acidity      1319 non-null   float64
 2   citric acid           1319 non-null   float64
 3   residual sugar        1319 non-null   float64
 4   chlorides             1319 non-null   float64
 5   free sulfur dioxide   1319 non-null   float64
 6   total sulfur dioxide  1319 non-null   float64
 7   density               1319 non-null   float64
 8   pH                    1319 non-null   float64
 9   sulphates             1319 non-null   float64
 10  alcohol               1319 non-null   float64
 11  quality               1319 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 134.0 KB


In [19]:
# 삭제된 결과 확개
df3.quality.value_counts()

5    681
6    638
Name: quality, dtype: int64

In [20]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1319 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1319 non-null   float64
 1   volatile acidity      1319 non-null   float64
 2   citric acid           1319 non-null   float64
 3   residual sugar        1319 non-null   float64
 4   chlorides             1319 non-null   float64
 5   free sulfur dioxide   1319 non-null   float64
 6   total sulfur dioxide  1319 non-null   float64
 7   density               1319 non-null   float64
 8   pH                    1319 non-null   float64
 9   sulphates             1319 non-null   float64
 10  alcohol               1319 non-null   float64
 11  quality               1319 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 134.0 KB


In [21]:
df3

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
0         9.4        5  
1         9.8        5  
2         9.8        5  
3         9.8        6  
4         9.4        5  
...       ...      ...  
1594     10.5        5  
1595     11.2        6  
1596     11.0        6  
1597     10.2        5  
1598     11.0        6  

[1319 rows x 12 columns]

In [22]:
# knn
ndf = df3

x = ndf[['fixed acidity', 'residual sugar', 'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density', 'pH', 'alcohol', 'sulphates']]
y = ndf.quality

x = preprocessing.StandardScaler().fit(x).transform(x)

Xtrain, Xtest, Ytrain, Ytest = train_test_split(x, y, test_size=0.9, random_state=26)

knn = KNeighborsClassifier(n_neighbors=100)

knn.fit(Xtrain, Ytrain.values.ravel())

Yhat = knn.predict(Xtest)
print(Yhat[0:10])
print(Ytest.values[0:10])

[5 5 5 5 5 5 5 5 5 5]
[5 6 6 5 6 6 5 6 5 6]


In [23]:
# Knn - Confusion_matrix
knnMatrix = metrics.confusion_matrix(Ytest, Yhat)
print(knnMatrix)

knnReport = metrics.classification_report(Ytest, Yhat)
print(knnReport)

[[570  44]
 [419 155]]
              precision    recall  f1-score   support

           5       0.58      0.93      0.71       614
           6       0.78      0.27      0.40       574

    accuracy                           0.61      1188
   macro avg       0.68      0.60      0.56      1188
weighted avg       0.67      0.61      0.56      1188



ㅌ## knn 결과 해석
- 5번으로의 예측 정확도 : 0.71
- 6번으로의 예측 정확도 : 0.69
- 예측 정확도가 낮다

In [ ]:
# Decision Tree - quality 예측하기

X = df3[['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar', 'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density', 'pH', 'sulphates', 'alcohol']]

y = df3['quality']

X = preprocessing.StandardScaler().fit(X).transform(X)

xTrain, xTest, yTrain, yTest = train_test_split(X, y, test_size=0.1, random_state=5)

treeModel = tree.DecisionTreeClassifier(criterion= 'entropy', max_depth= 5)

treeModel.fit(xTrain, yTrain)

yHat = treeModel.predict(xTest)

print(yHat[0:10])
print(yTest.values[0:10])

[5 5 6 6 6 6 6 5 5 5]
[5 5 5 5 6 6 6 5 5 6]


In [ ]:
# Decision Tree - Confusion_matrix
DTMatrix = metrics.confusion_matrix(yTest, yHat)
print(DTMatrix)

DFReport = metrics.classification_report(yTest, yHat)
print(DFReport)

[[33 29]
 [14 56]]
              precision    recall  f1-score   support

           5       0.70      0.53      0.61        62
           6       0.66      0.80      0.72        70

    accuracy                           0.67       132
   macro avg       0.68      0.67      0.66       132
weighted avg       0.68      0.67      0.67       132



## Decision Tree 결과 해석
- 5번으로의 예측 정확도 : 0.66
- 6번으로의 예측 정확도 : 0.69
- 예측 정확도가 낮다
- KNN이 Decision Tree보다 분류 결과가 조금 더 정확하다

In [34]:
# 5,6 뺀거

ddf = df2

x = ddf[['fixed acidity', 'residual sugar', 'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density', 'pH', 'alcohol', 'sulphates']]
y = ddf.quality

x = preprocessing.StandardScaler().fit(x).transform(x)

Xtrain, Xtest, Ytrain, Ytest = train_test_split(x, y, test_size=0.9, random_state=26)

knn = KNeighborsClassifier(n_neighbors=10)

knn.fit(Xtrain, Ytrain.values.ravel())

Yhat = knn.predict(Xtest)
print(Yhat[0:10])
print(Ytest.values[0:10])

[5 6 5 5 5 6 6 5 6 5]
[6 6 5 6 5 6 6 5 6 5]


In [35]:
# Knn - Confusion_matrix
knnMatrix = metrics.confusion_matrix(Ytest, Yhat)
print(knnMatrix)

knnReport = metrics.classification_report(Ytest, Yhat)
print(knnReport)

[[  0   0   6   3   0   0]
 [  0   0  29  19   1   0]
 [  0   0 499 104   5   0]
 [  0   0 277 284  16   0]
 [  0   0  36 127  18   0]
 [  0   0   1  10   5   0]]
              precision    recall  f1-score   support

           3       0.00      0.00      0.00         9
           4       0.00      0.00      0.00        49
           5       0.59      0.82      0.69       608
           6       0.52      0.49      0.51       577
           7       0.40      0.10      0.16       181
           8       0.00      0.00      0.00        16

    accuracy                           0.56      1440
   macro avg       0.25      0.24      0.23      1440
weighted avg       0.51      0.56      0.51      1440



/opt/anaconda3/envs/data_science/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/envs/data_science/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/envs/data_science/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
# 5,6 뺀거

ddf = df2

x = ndf[['fixed acidity', 'residual sugar', 'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density', 'pH', 'alcohol', 'sulphates']]
y = ndf.quality

x = preprocessing.StandardScaler().fit(x).transform(x)

Xtrain, Xtest, Ytrain, Ytest = train_test_split(x, y, test_size=0.9, random_state=26)

knn = KNeighborsClassifier(n_neighbors=100)

knn.fit(Xtrain, Ytrain.values.ravel())

Yhat = knn.predict(Xtest)
print(Yhat[0:10])
print(Ytest.values[0:10])

NameError: name 'df2' is not defined